### ISSUE TITLE
*log.safeError: replica.go:1724: on-disk and in-memory state diverged: [Lease.Expiration: nil != &hlc.Timestamp{}...

### ISSUE THREAD

**Issue by danhhz** [cockroachdb/cockroach#19843]
https://sentry.io/cockroach-labs/cockroachdb/issues/400115795/

```
*log.safeError: replica.go:1724: on-disk and in-memory state diverged: [Lease.Expiration: nil != &hlc.Timestamp{} Lease.DeprecatedStartStasis: nil != &hlc.Timestamp{}]
  File "github.com/cockroachdb/cockroach/pkg/storage/replica_raftstorage.go", line 851, in applySnapshot
  File "github.com/cockroachdb/cockroach/pkg/storage/store.go", line 3227, in func1
  File "github.com/cockroachdb/cockroach/pkg/storage/store.go", line 2970, in withReplicaForRequest
  File "github.com/cockroachdb/cockroach/pkg/storage/store.go", line 3036, in processRaftSnapshotRequest
  File "github.com/cockroachdb/cockroach/pkg/storage/store.go", line 2839, in HandleSnapshot
...
(3 additional frame(s) were not displayed)

replica.go:1724: on-disk and in-memory state diverged: [Lease.Expiration: nil != &hlc.Timestamp{} Lease.DeprecatedStartStasis: nil != &hlc.Timestamp{}]
```

**Comment by petermattis** [comment]
This was for one of our clusters, right? Can you point me to the cluster?

**Comment by danhhz** [comment]
This happened while doing a rolling upgrade of adriatic node 1 to current master. I think it was previously on 1.1.2, diving the logs to make sure

**Comment by danhhz** [comment]
If my grep-ing is right, the last version it started up with before this was `CockroachDB CCL v1.1.1-46-gc726df2 (linux amd64, built 2017/10/26 22:27:17, go1.8.3)`

**Comment by petermattis** [comment]
`v1.1.1-46-gc726df2` is also known as `v1.1.2`. And the new version (which crashed) is `CockroachDB CCL 3b90c52 (linux amd64, built 2017/11/06 17:16:45, go1.9)`.

**Comment by petermattis** [comment]
Looks like another problem with making those fields nullable. Consider this test (where `DeprecatedLease` is the old lease proto with `Expiration` and `DeprecatedStartStasis` marked non-nullable):

```
func TestZeroLeaseMarshalRoundTrip(t *testing.T) {
	b, err := protoutil.Marshal(&DeprecatedLease{})
	if err != nil {
		t.Fatal(err)
	}

	l := &Lease{}
	if err := protoutil.Unmarshal(b, l); err != nil {
		t.Fatal(err)
	}

	fmt.Printf("Expiration: %p\n", l.Expiration)
	fmt.Printf("DeprecatedStartStasis: %p\n", l.DeprecatedStartStasis)
}
```

This outputs:

```
Expiration: 0xc420014f10
DeprecatedStartStasis: 0xc420014f20
```

I could have sworn this was handled in the code, but I'm failing to see where.

**Comment by petermattis** [comment]
The assertion failed when checking disk state vs in-memory state after applying a snapshot. I'm still piecing together the code path here, but a node running `v1.1.2` sent a preemptive snapshot to a node run `v1.2-alpha`. The snapshot header contains the `ReplicaState` at the time the snapshot was taken. We then check that this state is equal to the on disk state after applying the snapshot. I'm not seeing how the problem mentioned in the previous comment would cause a problem. In both cases, we're decoding a `Lease` that was marshalled by `v1.1.2` in a `v1.2-alpha` binary. `Lease.Expiration` and `Lease.DeprecatedStartStasis` should be `&hlc.Timestamp{}`. 

I'm probably just missing something obvious.

**Comment by petermattis** [comment]
Looks like we have the following scenario:
1. 10/20 Deploy v1.2-alpha.20170901-969-g4b60590
1. 10/27 Downgrade to v1.1.1-46-gc726df2
1. 11/06 Deploy 3b90c52 (v1.2-alpha...)

What I think happened is that while running `v1.2-alpha.20170901-969-g4b60590`, we acquired the lease for `r1756`. We then downgraded to `v1.1.1-46-gc726df2` but the lease was never reacquired and thus never overwrote the on-disk state. When `3b90c52` was deployed today, a preemptive snapshot was sent from a `v1.1` node to the `v1.2` node and the snapshot header contained a lease containing a "zero" expiration while the on-disk state contained a lease with no expiration field. Boom.

I need to think about what can be done to prevent this. Perhaps just clearing the `Lease.{Expiration,DeprecatedStartStasis}` fields when loading the lease. I don't think this should block this week's alpha as the problem isn't new.

**Comment by danhhz** [comment]
@petermattis are you ready for me to bring node 1 back? do i need to wipe it or should restarting it be sufficient?

**Comment by petermattis** [comment]
Restarting the node should be sufficient.

**Comment by petermattis** [comment]
Clearing `Lease.{Expiration,DeprecatedStartStasis}` when loading the lease would not be sufficient. We could add a `Lease.Normalize()` method clears those fields for epoch-based leases and do that whenever the in-memory lease state is set.

Another option is to change `hlc.Timestamp.Equal` to consider `nil` and the "zero" timestamp as equal. This would avoid whack-a-mole, but is mildly surprising. 

@tschottdorf, @bdarnell thoughts?

**Comment by tschottdorf** [comment]
You're suggesting "sanitizing" the lease to match the on-disk version, right? I think we have to still look at the disk (perhaps you implied this) because whenever you (at 1.2) receive a state with a lease with a zero (instead of nil) field, you don't know whether the associated `WriteBatch` has `nil` or `zero` for the field (`zero` in the common case, `nil` in this issue).

But if the (in-memory update) is `nil` it's always ok (because 1.2 won't send out a `nil` field unless the `WriteBatch` also has `nil`?).

So what I'm thinking is that before any update of `r.mu.lease` with a zero field, you read from the engine and if the field is `nil` there, that takes precedence.

**Comment by petermattis** [comment]
I'm suggesting "sanitizing" the lease before comparison. Something like:

```
func (l *Lease) Normalize() {
	l.DeprecatedStartStasis = nil
	switch l.Type() {
	case LeaseEpoch:
		l.Expiration = nil
	}
}
```

It might be sufficient to call `Lease.Normalize()` on both the disk and in-memory leases in `Replica.assertStateLocked()`.

**Comment by bdarnell** [comment]
> Another option is to change hlc.Timestamp.Equal to consider nil and the "zero" timestamp as equal. This would avoid whack-a-mole, but is mildly surprising.

I think I'd do this, but as narrowly as possible, on Lease.Equal instead of hlc.Timestamp. 

Now that data.proto is proto3, can we change these two fields to be non-nullable again? Or does proto3's "don't encode zero values" rule only apply to primitive fields?

**Comment by petermattis** [comment]
> I think I'd do this, but as narrowly as possible, on Lease.Equal instead of hlc.Timestamp.

I'll give this a shot.

> Now that data.proto is proto3, can we change these two fields to be non-nullable again? Or does proto3's "don't encode zero values" rule only apply to primitive fields?

It only applies to primitive fields. Non-primitive fields are nullable by default. And if we mark those fields non-nullable, we're back to encoding the zero value again.

**Comment by tschottdorf** [comment]
> I'm suggesting "sanitizing" the lease before comparison. Something like:

How does this solve the problem? A node could be sending you a lease that you sanitize to a `nil` state, but that really has a zero field on-disk, and then you'd still be divergent.

**Comment by petermattis** [comment]
> How does this solve the problem? A node could be sending you a lease that you sanitize to a nil state, but that really has a zero field on-disk, and then you'd still be divergent.

You'd have to sanitize both the on-disk and in-memory leases before comparison. #19867 is essentially this approach, but doing so via changing `Lease.Equal`.

**Comment by tschottdorf** [comment]
Ah, it wasn't clear to me that both would be sanitized. SGTM though I'd personally prefer an approach that avoids the divergence instead of hides it (due to our past experience with other code that needed to be aware of the divergence to function properly). I'm ok with #19867 though.

**Comment by petermattis** [comment]
This was fixed by #19867.
